src/lib/pipeline/DispatchIssueUnit.cc

```diff
@@ -67,31 +67,42 @@ void DispatchIssueUnit::tick() {
       continue;
     }
 
-    const std::vector<uint16_t>& supportedPorts = uop->getSupportedPorts();
+    std::vector<uint16_t> supportedPorts = uop->getSupportedPorts();
     if (uop->exceptionEncountered()) {
       // Exception; mark as ready to commit, and remove from pipeline
       uop->setCommitReady();
       input_.getHeadSlots()[slot] = nullptr;
       continue;
     }
-    // Allocate issue port to uop
-    uint16_t port = portAllocator_.allocate(supportedPorts);
-    uint16_t RS_Index = portMapping_[port].first;
-    uint16_t RS_Port = portMapping_[port].second;
-    assert(RS_Index < reservationStations_.size() &&
-           "Allocated port inaccessible");
-    ReservationStation& rs = reservationStations_[RS_Index];
 
-    // When appropriate, stall uop or input buffer if stall buffer full
-    if (rs.currentSize == rs.capacity ||
-        dispatches_[RS_Index] == rs.dispatchRate) {
-      // Deallocate port given
-      portAllocator_.deallocate(port);
+    // Loop through all ports and remove any who's RS is at capacity or dispatch
+    // rate has been met
+    auto portIt = supportedPorts.begin();
+    while (portIt != supportedPorts.end()) {
+      uint16_t RS_Index = portMapping_[*portIt].first;
+      ReservationStation* rs = &reservationStations_[RS_Index];
+      if (rs->currentSize == rs->capacity ||
+          dispatches_[RS_Index] == rs->dispatchRate) {
+        portIt = supportedPorts.erase(portIt);
```

Comment:
cpp reference says "Note This operation could be expensive and if it is frequently used the user should consider using std::list."

Reply:
Presume this is about the erase?

Reply:
Ah yeah, sorry, should have clarified. This is regarding the erase

Reply:
So using goldbolt I timed 1,000,000 loops of: creating a vector/list of 10 elements and removing all elements one by one in a while loop using an iterator and `erase`. 

The results don't differ regardless of optimisation flag: vector is always quicker. I presume due to contiguous memory layout. Platform used was x86_64 with gcc 14.2:

- No optimisation flags, vector is quicker (2,730,431ms vs. 3,921,068ms) https://godbolt.org/z/77sMTfxax
- `-O2`, vec is quicker (47,576ms vs. 236,742ms) https://godbolt.org/z/afcc9ae5f
- `-O3`, vec is quicker (98,133ms vs. 738,515ms) https://godbolt.org/z/c5s16qbo6

Reply:
Very thorough. It is a simple example but the reference doesn't give any more information as to what cases are slow. Seems like vec is a lot better in your example. Happy to leave as is